core/src/validator.rs

```diff
@@ -1925,85 +1929,106 @@ fn maybe_warp_slot(
-                if let Ok(ref backup_blockstore) = backup_blockstore {
-                    copied += shreds.len();
-                    let _ = backup_blockstore.insert_shreds(shreds, None, true);
-                }
-            }
-            if last_print.elapsed().as_millis() > 3000 {
-                info!(
-                    "Copying shreds from slot {} copied {} so far.",
-                    start_slot, copied
-                );
-                last_print = Instant::now();
-            }
-            last_slot = Some(slot);
+    // .unwrap() safe because getting to this point implies blockstore has slots/shreds
+    let end_slot = blockstore.highest_slot()?.unwrap();
+
+    // Backing up the shreds that will be deleted from primary blockstore is
+    // not critical, so swallow errors from backup blockstore operations.
+    let backup_folder = format!(
+        "{}_backup_{}_{}_{}",
+        config
+            .ledger_column_options
+            .shred_storage_type
+            .blockstore_directory(),
+        timestamp() / 1000,
+        start_slot,
+        end_slot
+    );
+    match Blockstore::open_with_options(
+        &ledger_path.join(backup_folder),
+        blockstore_options_from_config(config),
+    ) {
+        Ok(backup_blockstore) => {
+            info!("Backing up slots from {start_slot} to {end_slot}");
+            let (_, backup_time) = measure!(
+                {
+                    let last_print = AtomicInterval::default();
+                    let mut num_slots_copied = 0;
+                    let slot_meta_iterator = blockstore.slot_meta_iterator(start_slot)?;
+                    for (slot, _meta) in slot_meta_iterator {
+                        let shreds = blockstore.get_data_shreds_for_slot(slot, 0)?;
+                        let _ = backup_blockstore.insert_shreds(shreds, None, true);
+                        num_slots_copied += 1;
+
+                        if last_print.should_update(5000) {
+                            info!("Copied {num_slots_copied} slots");
+                        }
+                    }
+                },
+                "blockstore backup"
+            );
+            info!("done. {backup_time}");
+        }
+        Err(err) => {
+            warn!("Unable to backup shreds with incorrect shred version: {err}");
         }
-
-        let end_slot = last_slot.unwrap();
-        info!("Purging slots {} to {}", start_slot, end_slot);
-        blockstore.purge_from_next_slots(start_slot, end_slot);
-        blockstore.purge_slots(start_slot, end_slot, PurgeType::Exact);
-        info!("done");
     }
-    drop(blockstore);
+
+    info!("Purging slots {start_slot} to {end_slot} from blockstore");
+    let (_, purge_time) = measure!(
+        {
+            blockstore.purge_from_next_slots(start_slot, end_slot);
+            blockstore.purge_slots(start_slot, end_slot, PurgeType::Exact);
+        },
+        "blockstore purge"
+    );
+    info!("done. {purge_time}");
```

Comment:
there are other threads spewing logs between these, right?  always annoys me to see a solitary "done" log line. a consistent marker string on the start and end logs would be great